transport/transporters/rabbitmq/transporter/transporter.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	defer func() {
+		// recover if channel is already closed
+		recover()
+	}()
+
+	t.log.Debug("closing progress channel")
+	close(t.txnsWritten)
+}
+
+// StartTransporting reads in message batches, outputs its String to RabbitMQ and then sends a progress report on the batch
+func (t RabbitMQTransporter) StartTransporting() {
+	t.log.Info("starting transporter")
+	defer t.shutdown()
+
+	var b interface{}
+	var ok bool
+	var err error
+
+	t.channel, err = t.conn.Channel()
+	if err != nil {
+		t.log.WithError(err).Fatal("Could not open channel to RabbitMQ")
+	}
+	defer func() {
+		err = t.channel.Close()
+		if err != nil {
+			t.log.WithError(err).Error("Error while closing channel")
+		}
+	}()
+	closeNotify := t.channel.NotifyClose(make(chan wabbit.Error))
+	t.publishNotify = t.channel.NotifyPublish(make(chan wabbit.Confirmation, t.batchSize))
+	err = t.channel.Confirm(false)
+	if err != nil {
+		t.log.WithError(err).Fatal("Could not turn on confirmations for channel")
+	}
+	confirms := uint64(0)
+
+	for {
+		select {
+		case <-t.shutdownHandler.TerminateCtx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			return
+
+		case b, ok = <-t.inputChan:
+			// pass
+		}
+
+		select {
+		case <-t.shutdownHandler.TerminateCtx.Done():
+			t.log.Debug("received terminateCtx cancellation")
+			return
+		case err := <-closeNotify:
+			t.log.Error(err.Error())
+			return
+		default:
+			// pass
+		}
+
+		if !ok {
+			t.log.Warn("input channel is closed")
+			return
+		}
+
+		genericBatch, ok := b.(*batch.GenericBatch)
+
+		if !ok {
+			panic("Batch is not a GenericBatch")
+		}
+
+		messages := genericBatch.GetPayload()
+		messagesSlice, ok := messages.([]*marshaller.MarshalledMessage)
+
+		if !ok {
+			panic("Batch payload is not a []*marshaller.MarshalledMessage")
+		}
+
+		// Begin timer
+		start := TimeSource.UnixNano()
+
+		err = t.sendMessages(messagesSlice)
+		if err != nil {
+			t.log.WithError(err).Error("Could not transport messages")
+			return
+		}
+
+		confirms, err = t.waitForConfirmations(uint64(len(messagesSlice)), confirms)
+		if err != nil {
+			t.log.WithError(err).Error("Could not confirm message publication")
+			return
+		}
+
+		// End timer and send stat
+		total := (TimeSource.UnixNano() - start) / int64(time.Millisecond)
+		t.statsChan <- stats.NewStatHistogram("rabbitmq_transport", "duration", total, TimeSource.UnixNano(), "ms")
+
+		t.log.Debug("Messages delivered")
+		t.statsChan <- stats.NewStatCount("rabbitmq_transport", "written", int64(len(messagesSlice)), TimeSource.UnixNano())
+
+		// report transactions written in this batch
+		t.txnsWritten <- genericBatch.GetTransactions()
+	}
+}
+
+type operation struct {
+	Operation string `json:"operation"`
+	Table     string `json:"table"`
+}
+
+func (t RabbitMQTransporter) sendMessages(messagesSlice []*marshaller.MarshalledMessage) error {
+	var err error
+	op := operation{}
+	for _, msg := range messagesSlice {
+		err = json.Unmarshal(msg.Json, &op)
```

Comment:
`MarshalledMessage` already has `operation`. It's reasonable to add `table` too so you don't have to unmarshal here.

Reply:
Oh, perfect.  Thank you.